### PR TITLE
Client options and fix sub issue

### PR DESCRIPTION
**Pull request by bquantump**

### CODE REVIEW

File: azure-proto-core/ArmClient.cs [hunk 1]
```diff
@@ -24,23 +24,23 @@ public class ArmClient
...
 
-        public ArmClient(string defaultSubscriptionId) : this(new Uri(DefaultUri), new DefaultAzureCredential(), defaultSubscriptionId, new ArmClientOptions()) { }
+        public ArmClient(string defaultSubscriptionId, ArmClientOptions options = null) : this(new Uri(DefaultUri), new DefaultAzureCredential(), defaultSubscriptionId, options == null ? new ArmClientOptions() : options) { }
 
-        public ArmClient(TokenCredential credential, string defaultSubscriptionId) : this(new Uri(DefaultUri), credential, defaultSubscriptionId, new ArmClientOptions()) { }
+        public ArmClient(TokenCredential credential, string defaultSubscriptionId, ArmClientOptions options = null) : this(new Uri(DefaultUri), credential, defaultSubscriptionId, options == null ? new ArmClientOptions() : options) { }
 
-        public ArmClient(Uri baseUri, TokenCredential credential) : this(baseUri, credential, null, new ArmClientOptions()) { }
+        public ArmClient(Uri baseUri, TokenCredential credential, ArmClientOptions options = null) : this(baseUri, credential, null, options == null ? new ArmClientOptions() : options) { }
```
Comment (m-nash) [inline azure-proto-core/ArmClient.cs:33]: You can use null coalesce here instead

opitions ?? new ArmClientOptions()

File: azure-proto-core/ArmClientOptions.cs [hunk 2]
```diff
@@ -1,14 +1,34 @@
...
 {
-    public class ArmClientOptions
+    public class ArmClientOptions : ClientOptions
     {
         private Dictionary<Type, object> _overrides = new Dictionary<Type, object>();
         private static readonly object _overridesLock = new object();
 
+        //TODO policy lists are internal hence we don't have acces to them by inheriting ClientOptions in this Asembly, this is a wrapper for now to convert to the concrete
+        //policy options. 
+        public new void AddPolicy(HttpPipelinePolicy policy, HttpPipelinePosition position){
```
Comment (m-nash) [inline azure-proto-core/ArmClientOptions.cs:15]: Why is this necessary?  AddPolicy(...) is public so we shouldn't need access to the direct lists themselves do we?

File: azure-proto-core/OperationsBase.cs [hunk 3]
```diff
@@ -31,9 +32,11 @@ public OperationsBase(ArmClientContext context, Resource resource)
 
         public virtual ResourceIdentifier Id { get; }
 
+        public virtual ArmClientOptions ClientOptions {get; protected set;}
+
         public virtual void Validate(ResourceIdentifier identifier)
         {
-            if (identifier?.Type != ResourceType)
+            if (SubscriptionContainerOperations.AzureResourceType != identifier && identifier?.Type != ResourceType)
```
Comment (m-nash) [inline azure-proto-core/OperationsBase.cs:39]: I think this should be

```suggestion
            if (SubscriptionContainerOperations.AzureResourceType != ResourceType && identifier == null)
```

File: azure-proto-core/OperationsBase.cs [hunk 4]
```diff
@@ -8,9 +8,9 @@ namespace azure_proto_core
     /// </summary>
     public abstract class OperationsBase
     {
-        public OperationsBase(ArmClientContext context, ResourceIdentifier id, Location location = null) : this(context, new ArmResource(id, location ?? Location.Default)) { }
+        public OperationsBase(ArmClientContext context, ResourceIdentifier id, Location location = null, ArmClientOptions ClientOptions = null) : this(context, new ArmResource(id, location ?? Location.Default), ClientOptions) { }
```
Comment (m-nash) [inline azure-proto-core/OperationsBase.cs:11]: shouldn't this be required?

File: azure-proto-core/Resources/ResourceIdentifier.cs [hunk 5]
```diff
@@ -28,7 +28,7 @@ public ResourceIdentifier(string id)
         {
             Id = id;
             //TODO: Due to the implicit this is called for blank constructions such as new PhResourceGroup
-            if (id != null)
+            if (SubscriptionContainerOperations.AzureResourceType != id && id != null)
```
Comment (m-nash) [inline azure-proto-core/Resources/ResourceIdentifier.cs:31]: this shouldn't be needed with the other changes

File: azure-proto-core/SubscriptionContainerOperations.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+namespace azure_proto_core
+{
+    /// <summary>
+    /// Subscription Container Operationss
+    /// </summary>
+    public class SubscriptionContainerOperations : OperationsBase
+    {
+        public static readonly string AzureResourceType = "Microsoft.Resources/subscriptions";
+
+        public SubscriptionContainerOperations(ArmClientContext context, ArmClientOptions options) : base(context, AzureResourceType, null, options) { }
```
Comment (m-nash) [inline azure-proto-core/SubscriptionContainerOperations.cs:19]: I am wondering whether context and options should be separate classes?

**Review by m-nash — CHANGES_REQUESTED** [review]

File: azure-proto-core/ArmClientOptions.cs (hunk 2 above)
Comment (bquantump) [reply azure-proto-core/ArmClientOptions.cs:15]: Because we need to access list to copy over the polices for the concrete types

https://github.com/m-nash/azure-proto-sdk/blob/de6bee33ea5f092501d7283359db69d214ce5a51/azure-proto-core/SubscriptionContainerOperations.cs#L58

File: azure-proto-core/OperationsBase.cs (hunk 3 above)
Comment (bquantump) [reply azure-proto-core/OperationsBase.cs:39]: I don't think this would work. 

Just tried it real quick with your change suggestion and the subsequent this shouldn't be needed with the other changes

Time Elapsed 00:00:03.36
Unhandled exception. System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter ''Microsoft.Resources/subscriptions' is not a valid resource')
   at azure_proto_core.ResourceIdentifier.Parse(String id) in /mnt/c/Users/stevens/projects/dotnetsdk/azure-proto-sdk/azure-proto-core/Resources/ResourceIdentifier.cs:line 130
   at azure_proto_core.ResourceIdentifier..ctor(String id) in /mnt/c/Users/stevens/projects/dotnetsdk/azure-proto-sdk/azure-proto-core/Resources/ResourceIdentifier.cs:line 32
   at azure_proto_core.ResourceIdentifier.op_Implicit(String other) in /mnt/c/Users/stevens/projects/dotnetsdk/azure-proto-sdk/azure-proto-core/Resources/ResourceIdentifier.cs:line 199
   at azure_proto_core.SubscriptionContainerOperations..ctor(ArmClientContext context, ArmClientOptions options) in /mnt/c/Users/stevens/projects/dotnetsdk/azure-proto-sdk/azure-proto-core/SubscriptionContainerOperations.cs:line 19
   at azure_proto_core.ArmClient.Subscriptions() in /mnt/c/Users/stevens/projects/dotnetsdk/azure-proto-sdk/azure-proto-core/ArmClient.cs:line 62
   at azure_proto_core.ArmClient.GetDefaultSubscription(CancellationToken token) in /mnt/c/Users/stevens/projects/dotnetsdk/azure-proto-sdk/azure-proto-core/ArmClient.cs:line 197
   at azure_proto_core.ArmClient..ctor(Uri baseUri, TokenCredential credential, String defaultSubscriptionId, ArmClientOptions options) in /mnt/c/Users/stevens/projects/dotnetsdk/azure-proto-sdk/azure-proto-core/ArmClient.cs:line 41
   at azure_proto_core.ArmClient..ctor(ArmClientOptions options) in /mnt/c/Users/stevens/projects/dotnetsdk/azure-proto-sdk/azure-proto-core/ArmClient.cs:line 27
   at client.ClientOptionsOverride.Execute() in /mnt/c/Users/stevens/projects/dotnetsdk/azure-proto-sdk/client/Scenarios/ClientOverrides.cs:line 23
   at client.Program.Main(String[] args) in /mnt/c/Users/stevens/projects/dotnetsdk/azure-proto-sdk/client/Program.cs:line 14

File: azure-proto-core/ResourceGroupContainerOperations.cs [hunk 7]
```diff
@@ -15,60 +15,60 @@ public class ResourceGroupContainerOperations : ResourceContainerOperations<XRes
...
 
         public AsyncPageable<XResourceGroup> ListAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
             return new PhWrappingAsyncPageable<ResourceGroup, XResourceGroup>(
                 Operations.ListAsync(null, null, cancellationToken),
-                s => new XResourceGroup(ClientContext, new PhResourceGroup(s)));
+                s => new XResourceGroup(ClientContext, new PhResourceGroup(s), this.ClientOptions));
         }
 
         internal ResourceGroupsOperations Operations => GetClient<ResourcesManagementClient>((uri, cred) => new ResourcesManagementClient(uri, Id.Subscription, cred)).ResourceGroups;
```
Comment (m-nash) [inline azure-proto-core/ResourceGroupContainerOperations.cs:74]: should forward the new options down

File: azure-proto-compute/VirtualMachineOperations.cs [hunk 8]
```diff
@@ -81,42 +81,42 @@ public override ArmResponse<XVirtualMachine> Get()
...
 
         public async Task<ArmOperation<XVirtualMachine>> AddTagAsync(string key, string value, CancellationToken cancellationToken = default)
         {
             var patchable = new VirtualMachineUpdate { Tags = new Dictionary<string, string>() };
             patchable.Tags.Add(key, value);
-            return new PhArmOperation<XVirtualMachine, VirtualMachine>(await Operations.StartUpdateAsync(Id.ResourceGroup, Id.Name, patchable, cancellationToken), v => { Resource = new PhVirtualMachine(v); return new XVirtualMachine(ClientContext, Resource as PhVirtualMachine); });
+            return new PhArmOperation<XVirtualMachine, VirtualMachine>(await Operations.StartUpdateAsync(Id.ResourceGroup, Id.Name, patchable, cancellationToken), v => { Resource = new PhVirtualMachine(v); return new XVirtualMachine(ClientContext, Resource as PhVirtualMachine, this.ClientOptions); });
         }
 
         internal VirtualMachinesOperations Operations => GetClient<ComputeManagementClient>((baseUri, creds) => new ComputeManagementClient(baseUri, Id.Subscription, creds)).VirtualMachines;
```
Comment (m-nash) [inline azure-proto-compute/VirtualMachineOperations.cs:122]: should use the new options

File: azure-proto-compute/XAvailabilitySet.cs [hunk 9]
```diff
@@ -4,7 +4,7 @@ namespace azure_proto_compute
 {
     public class XAvailabilitySet : AvailabilitySetOperations
     {
-        public XAvailabilitySet(ArmClientContext context, PhAvailabilitySet resource):base(context, resource)
+        public XAvailabilitySet(ArmClientContext context, PhAvailabilitySet resource, ArmClientOptions clientOptions): base(context, resource, clientOptions)
```
Comment (m-nash) [inline azure-proto-compute/XAvailabilitySet.cs:7]: not sure we want this here as this introduces the opportunity for people to "change" the opitions inside of a client instance

File: azure-proto-core/ResourceGroupOperations.cs [hunk 10]
```diff
@@ -29,26 +29,26 @@ public async Task<ArmOperation<Response>> DeleteAsync(CancellationToken cancella
...
 
         public async Task<ArmOperation<XResourceGroup>> AddTagAsync(string name, string value, CancellationToken cancellationToken = default)
         {
             var patch = new ResourceGroupPatchable();
             patch.Tags[name] = value;
-            return new PhArmOperation<XResourceGroup, ResourceGroup>(await Operations.UpdateAsync(Id.Name, patch, cancellationToken), g => { this.Resource = new PhResourceGroup(g); return new XResourceGroup(ClientContext, Resource as PhResourceGroup); });
+            return new PhArmOperation<XResourceGroup, ResourceGroup>(await Operations.UpdateAsync(Id.Name, patch, cancellationToken), g => { this.Resource = new PhResourceGroup(g); return new XResourceGroup(ClientContext, Resource as PhResourceGroup, this.ClientOptions); });
         }
 
         public ArmResponse<TOperations> CreateResource<TContainer, TOperations, TResource>(string name, TResource model, azure_proto_core.Location location = default)
```
Comment (m-nash) [inline azure-proto-core/ResourceGroupOperations.cs:54]: should forward the new opitions to the client

**Review by m-nash — CHANGES_REQUESTED** [review]

File: azure-proto-compute/AvailabilitySetContainer.cs [hunk 11]
```diff
@@ -14,9 +14,9 @@ namespace azure_proto_compute
     /// </summary>
     public class AvailabilitySetContainer : ResourceContainerOperations<XAvailabilitySet, PhAvailabilitySet>
     {
-        public AvailabilitySetContainer(ArmClientContext context, PhResourceGroup resourceGroup) : base(context, resourceGroup) { }
+        public AvailabilitySetContainer(ArmClientContext context, PhResourceGroup resourceGroup, ArmClientOptions clientOptions) : base(context, resourceGroup, clientOptions) { }
```
Comment (allenjzhang) [inline azure-proto-compute/AvailabilitySetContainer.cs:17]: Are we not merging Context and Options? Or is it to be done later.

Comment (bquantump) [reply azure-proto-compute/AvailabilitySetContainer.cs:17]: Later

File: azure-proto-compute/Extensions/AzureSubscriptionExtension.cs [hunk 12]
```diff
@@ -19,6 +19,7 @@ public static class AzureSubscriptionExtension
         /// <param name="top"></param>
         /// <param name="cancellationToken"></param>
         /// <returns></returns>
+
```
Comment (allenjzhang) [inline azure-proto-compute/Extensions/AzureSubscriptionExtension.cs:22]: nit. remove extra blank line

File: azure-proto-core/ArmClient.cs [hunk 13]
```diff
@@ -1,161 +1,109 @@
...
-        }
+        public ArmClient(string defaultSubscriptionId, ArmClientOptions options = null)
+            : this(new Uri(DefaultUri), new DefaultAzureCredential(), defaultSubscriptionId, options ?? new ArmClientOptions()) { }
 
-        public ArmClient(Uri baseUri, TokenCredential credential)
-            : this(baseUri, credential, null, new ArmClientOptions())
-        {
-        }
+        public ArmClient(TokenCredential credential, string defaultSubscriptionId, ArmClientOptions options = null)
+            : this(new Uri(DefaultUri), credential, defaultSubscriptionId, options ?? new ArmClientOptions()) { }
```
Comment (allenjzhang) [inline azure-proto-core/ArmClient.cs:34]: >options ?? new ArmClientOptions() [](start = 75, length = 33)

Don't need to have all these checks. Only need it at the lowest level of ctor and do it once.

File: azure-proto-core/ArmClient.cs [hunk 14]
```diff
@@ -1,161 +1,109 @@
...
+
+        public ArmClient(Uri baseUri, TokenCredential credential, string defaultSubscriptionId, ArmClientOptions options)
         {
+            ClientOptions = options;
             ClientContext = new ArmClientContext(baseUri, credential);
             defaultSubscriptionId ??= GetDefaultSubscription().ConfigureAwait(false).GetAwaiter().GetResult();
-            DefaultSubscription = new SubscriptionOperations(
-                ClientContext,
-                new ResourceIdentifier($"/subscriptions/{defaultSubscriptionId}"));
+            DefaultSubscription = new SubscriptionOperations(ClientContext, new ResourceIdentifier($"/subscriptions/{defaultSubscriptionId}"), options);
```
Comment (allenjzhang) [inline azure-proto-core/ArmClient.cs:46]: >new ResourceIdentifier [](start = 76, length = 22)

@m-nash, I feel there is an opportunity here to create some helper method for more structured ResourceIdentifier construction. There are still too much free typing going on here. Lets add a TODO for fit and finish.

File: azure-proto-core/ArmClient.cs [hunk 15]
```diff
@@ -167,143 +115,96 @@ await foreach (var location in ListAvailableLocationsAsync(subscriptionId, resou
...
+        /// Fill in the default subscription in the simple case (passed in, or only one subscription available)
         /// </summary>
         /// <param name="token"></param>
         /// <returns></returns>
-        internal async Task<string> GetDefaultSubscription(CancellationToken token = default)
+        internal async Task<string> GetDefaultSubscription(CancellationToken token = default(CancellationToken))
         {
-            var sub = DefaultSubscription?.Id?.Subscription;
+            string sub = DefaultSubscription?.Id?.Subscription;
             if (null == sub)
```
Comment (allenjzhang) [inline azure-proto-core/ArmClient.cs:199]: >null [](start = 16, length = 4)

string.IsNullOrWhitespace()?

Comment (bquantump) [reply azure-proto-core/ArmClient.cs:199]: Would not white space in this case be a case for an exception? If that happened, something went wrong.

File: azure-proto-core/ArmClientOptions.cs [hunk 16]
```diff
@@ -1,22 +1,54 @@
...
 using System.ComponentModel;
-
+using Azure.Core;
+using Azure.Core.Pipeline;
 namespace azure_proto_core
 {
-    public class ArmClientOptions
+    public class ArmClientOptions : ClientOptions
     {
+        public static T convert<T>(ArmClientOptions options) where T : ClientOptions, new()
```
Comment (m-nash) [inline azure-proto-core/ArmClientOptions.cs:10]: Methods should start with capital

File: azure-proto-core/SubscriptionContainerOperations.cs [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+namespace azure_proto_core
+{
+    /// <summary>
+    /// Subscription Container Operationss
+    /// </summary>
+    public class SubscriptionContainerOperations : OperationsBase
+    {
+        public static readonly string AzureResourceType = "Microsoft.Resources/subscriptions";
+
+        internal SubscriptionContainerOperations(ArmClientContext context, ArmClientOptions options) : base(context, null, options, null) { }
```
Comment (m-nash) [inline azure-proto-core/SubscriptionContainerOperations.cs:19]: base should be on a new line

**Review by m-nash — CHANGES_REQUESTED** [review]

**Review by m-nash — APPROVED** [review]
